arch/xtensa/include/xtensa-asm2-s.h

```diff
@@ -297,26 +319,43 @@ _restore_\@:
  * entry code (defined via EXCINT_HANDLER) and a C handler for this
  * particular level.
  *
- * FIXME: needs special handling for exceptions (level 1): it's "EPC"
- * and not "EPC1" (though IIRC the assembler makes this work).
- * And there is no EPS: instead PS is simply the interrupted PS
- * with EXCM flipped from 0 to 1.
- *
- * FIXME: needs better locking.  The hardware will NOT mask out "high
- * priority" exceptions on arrival here, so we have to do it ourselves
- * with RSIL.
+ * Note that the linker sections for some levels get special names for
+ * no particularly good reason.  Only level 1 has any code generation
+ * difference, because it is the legacy exception level that predates
+ * the EPS/EPC registers.
  */
 .macro DEF_EXCINT LVL, ENTRY_SYM, C_HANDLER_SYM
+.if \LVL == 1
+.pushsection .UserExceptionVector.text, "ax"
+.elseif \LVL == XCHAL_DEBUGLEVEL
+.pushsection .DebugExceptionVector.text, "ax"
+.elseif \LVL == XCHAL_NMILEVEL
+.pushsection .NMIExceptionVector.text, "ax"
+.else
 .pushsection .Level\LVL\()InterruptVector.text, "ax"
+.endif
 .global _Level\LVL\()Vector
 _Level\LVL\()Vector:
 	addi a1, a1, -BASE_SAVE_AREA_SIZE
 	s32i a0, a1, BSA_A0_OFF
 	s32i a2, a1, BSA_A2_OFF
 	s32i a3, a1, BSA_A3_OFF
 
+	/* Level "1" is the exception handler, which uses a different
+         * calling convention.  No special register holds the
+         * interrupted PS, instead we just assume that the CPU has
+         * turned on the EXCM bit and set INTLEVEL.
```

Comment:
This comment is wrong, "and set INTLEVEL" is actually not what it does (though it's what it *should* do, sigh).  Thus the "gotcha" comment and code above.  Will fix.